src/twisted/application/internet.py

```diff
@@ -523,13 +524,375 @@ def policy(attempt):
 
 
 
-def _noop():
+class _ClientMachine(object):
     """
-    Do nothing; this stands in for C{transport.loseConnection()} and
-    C{DelayedCall.cancel()} when L{ClientService} is in a state where there's
-    nothing to do.
+    State machine for maintaining a single outgoing connection to an
+    endpoint.
+
+    @see: L{ClientService}
     """
 
+    _machine = MethodicalMachine()
+
+    def __init__(self, endpoint, factory, retryPolicy, clock, log):
+        """
+        @see L{ClientService.__init__}
+
+        @param log: The logger for the L{ClientService} instance this
+            state machine is assoicated to.
+        @type log: L{Logger}
+        """
+        self._endpoint = endpoint
+        self._failedAttempts = 0
+        self._stopped = False
+        self._factory = factory
+        self._timeoutForAttempt = retryPolicy
+        self._clock = clock
+        self._connectionInProgress = succeed(None)
+
+        self._awaitingConnected = []
+
+        self._stopWaiters = []
+        self._log = log
+
+
+    @_machine.state(initial=True)
+    def _init(self):
+        """
+        The service has not been started.
+        """
+
+    @_machine.state()
+    def _connecting(self):
+        """
+        The service has started connecting.
+        """
+
+    @_machine.state()
+    def _waiting(self):
+        """
+        The service is waiting for the reconnection period
```

Comment:
Perhaps, but I'm not sure why you singled out this particular machine state? All of the states have some associated data that needs handling at transitions?